cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -0,0 +1,232 @@
+    byte[] prefix = keyPrefixForApplicationScan(appId);
+    try (Scanner scanner = store.scan(new Scan(prefix, Bytes.stopKeyForPrefix(prefix)))) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        store.delete(row.getRow());
+      }
+    }
+  }
+
+  /**
+   * Read a schedule from the store.
+   *
+   * @param scheduleId the id of the schedule to read
+   *
+   * @return the schedule from the store, or null if it was not found
+   */
+  @Nullable
+  public ProgramSchedule getSchedule(ScheduleId scheduleId) {
+    Row row = store.get(new Get(rowKeyForSchedule(scheduleId)));
+    byte[] serialized = row.get(SCHEDULE_COLUMN_BYTES);
+    return serialized == null ? null : GSON.fromJson(Bytes.toString(serialized), ProgramSchedule.class);
+  }
+
+  /**
+   * Retrieve all schedules for a given application.
+   *
+   * @param appId the application for which to list the schedules.
+   *
+   * @return a list of schedules for the application; never null
+   */
+  public List<ProgramSchedule> listSchedules(ApplicationId appId) {
+    return listSchedules(appId, null);
+  }
+
+  /**
+   * Retrieve all schedules for a given program.
+   *
+   * @param programId the program for which to list the schedules.
+   *
+   * @return a list of schedules for the program; never null
+   */
+  public List<ProgramSchedule> listSchedules(ProgramId programId) {
+    return listSchedules(programId.getParent(), programId);
+  }
+
+  /**
+   * Find all schedules that have a trigger with a given trigger key.
+   *
+   * @param triggerKey the trigger key to look up
+   *
+   * @return a list of all schedules that are triggered by this key; never null
+   */
+  public List<ProgramSchedule> findSchedules(String triggerKey) {
+    List<ProgramSchedule> result = new ArrayList<>();
+    try (Scanner scanner = store.readByIndex(TRIGGER_KEY_COLUMN_BYTES, Bytes.toBytes(triggerKey))) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        byte[] serialized = row.get(SCHEDULE_COLUMN_BYTES);
+        if (serialized != null) {
+          ProgramSchedule schedule = GSON.fromJson(Bytes.toString(serialized), ProgramSchedule.class);
+          result.add(schedule);
+        }
+      }
+    }
+    return result;
+  }
+
+  /*------------------- private helpers ---------------------*/
+
+  private List<ProgramSchedule> listSchedules(ApplicationId appId, @Nullable ProgramId programId) {
+    List<ProgramSchedule> result = new ArrayList<>();
+    byte[] prefix = keyPrefixForApplicationScan(appId);
+    try (Scanner scanner = store.scan(new Scan(prefix, Bytes.stopKeyForPrefix(prefix)))) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        byte[] serialized = row.get(SCHEDULE_COLUMN_BYTES);
+        if (serialized != null) {
+          ProgramSchedule schedule = GSON.fromJson(Bytes.toString(serialized), ProgramSchedule.class);
+          if (programId == null || programId.equals(schedule.getProgramId())) {
```

Comment:
could simplify to just `schedule.getProgramId().equals(programId)` since `schedule.getProgramId()` won't return `null`.